client/factory.go

```diff
@@ -0,0 +1,188 @@
+package client
+
+import (
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-07-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-05-01/resources"
+	"github.com/Azure/azure-sdk-for-go/services/storage/mgmt/2019-04-01/storage"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/k8sclient"
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/azure-operator/v4/pkg/credential"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+type FactoryConfig struct {
+	K8sClient  k8sclient.Interface
+	GSTenantID string
+}
+
+// Factory is creating Azure clients for specified AzureConfig CRs, so basically for specified
+// tenant clusters. All created clients are cached.
+type Factory struct {
+	mutex      sync.Mutex
+	k8sClient  k8sclient.Interface
+	gsTenantID string
+
+	clients map[string]*AzureClientSet
```

Comment:
Map of all clients in format `org's credential id` => `AzureClientSet`.

Reply:
We can create a new type in the package like `type credentialID string` so the map is more explicit `clients map[credentialID]*AzureClientSet`